USoC'23: Update session 04 - Binary Compatibility (#262)
unikraft/docs: content/en/community/hackathons/sessions/bincompat/content/Work-Items_02.-Build-app-elfloader.md

Using `./run*.sh`, we used the pre-built `app-elfloader` images from [the `run-app-elfloader` repository](https://github.com/unikraft/run-app-elfloader).
Let's now also build `app-elfloader`.
Building our custom `app-elfloader` image will be useful in case we added some changes inside Unikraft, and we want to use an `elfloader` image with those changes applied.

In order to build our own `elfloader` image, follow the instructions in the [`app-elfloader` README file](https://github.com/unikraft/app-elfloader#readme), up to the building section.
The build will result in an `elfloader_qemu-x86_64` kernel image.
Ignore the running instructions, we will do that by using the [`run.sh`](https://github.com/unikraft/run-app-elfloader/blob/master/run.sh) helper script.
The `run.sh` script can use a custom `elfloader` image with the `-k` option.

We will run the dynamic versions of `c++/helloworld`, the one in [the `dynamic-apps` repository](https://github.com/unikraft/dynamic-apps).
We will use the `-r` option to pass the root filesystem to the `run.sh` script:

```console
cd run-app-elfloader/

> ./run.sh -k ../app-elfloader/build/elfloader_qemu-x86_64 -r ../dynamic-apps/lang/c++/ "/helloworld"

Reviewer: ```suggestion
./run.sh -k ../app-elfloader/build/elfloader_qemu-x86_64 -r ../dynamic-apps/lang/c++/ /helloworld
```

```

This will use our `app-elfloader` image to run the C++ `helloworld` application.

```text
SeaBIOS (version rel-1.16.2-0-gea1b7a073390-prebuilt.qemu.org)
Booting from ROM..Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
                  Atlas 0.13.1~21ce1acf
Hello World!
```

Let's do the same for the `sqlite3` dynamically linked application:

```console
./run.sh -k ../app-elfloader/build/elfloader_qemu-x86_64 -r ../dynamic-apps/sqlite3/ "/usr/bin/sqlite3"
```

> This will give us the `sqlite3` cli prompt:

Reviewer: ```suggestion
This will give us the `sqlite3` CLI prompt:
```

> Using `./run*.sh` we ran the prebuilt images from [the `run-app-elfloader` repository](https://github.com/unikraft/run-app-elfloader).

Reviewer: ```suggestion
Using `./run*.sh`, we used the pre-built `app-elfloader` images from [the `run-app-elfloader` repository](https://github.com/unikraft/run-app-elfloader).
```

> Let's now also build `app-elfloader`.

Reviewer: Argue why this would be useful. Typically, it won't be; users just care about running applications. But, in case you just did a fix to Unikraft, and you want an `elfloader` image with that fix, you need to build it.

Run the following commands, similar to what we did in previous sessions when building usual applications:

```console

> $ git clone https://github.com/unikraft/app-elfloader/

Reviewer: Remove dollar sign.

I would add these to the `app-elfloader` repository as part of its README. And simply reference that README, without duplicating the information. We would amend [this PR](https://github.com/unikraft/app-elfloader/pull/22/files).

$ git clone https://github.com/unikraft/unikraft .unikraft/unikraft
$ git clone https://github.com/unikraft/lib-lwip .unikraft/libs/lwip
$ git clone https://github.com/unikraft/lib-libelf .unikraft/libs/libelf

$ UK_DEFCONFIG=$(pwd)/.config.elfloader_qemu-x86_64 make defconfig
/usr/bin/gcc -ldl -I. -I/tmp/usoc/app-elfloader/build/kconfig -DCONFIG_=\"\"   -c fixdep.c -o /tmp/usoc/app-elfloader/build/kconfig/fixdep.o
/usr/bin/gcc -ldl -I. -I/tmp/usoc/app-elfloader/build/kconfig -DCONFIG_=\"\"   /tmp/usoc/app-elfloader/build/kconfig/fixdep.o -o /tmp/usoc/app-elfloader/build/kconfig/fixdep
#
# configuration written to /tmp/usoc/app-elfloader/.config
#
make[1]: Leaving directory '/tmp/usoc/app-elfloader/.unikraft/unikraft'

$ make -j $(nproc)
[...]
  UKBI    elfloader_qemu-x86_64.dbg.bootinfo
  SCSTRIP elfloader_qemu-x86_64
  GZ      elfloader_qemu-x86_64.gz
make[1]: Leaving directory '/tmp/usoc/app-elfloader/.unikraft/unikraft'
```

This will result, as we already know, in a `app-elfloader_qemu-x86_64` kernel image.
To run an application using this `elfloader` image, we can still use the scripts from `run-app-elfloader`.

> The `run.sh` script can use a custom `elfloader` image with the `-k` option.

Reviewer: Add the `-k` command here directly. Assume a good directory hierarchy.

> Note that at the beginning of the sessions, we cloned the `app-elfloader`, `run-app-elfloader`, `dynamic-apps` and `static-pie-apps` repositories in the same root directory:

Reviewer: I would remove this, it's a bit confusing and I think it makes sense to consider the filesystem hierarchy already being set up.

```console
$ tree -L 1 .
.
|-- app-elfloader/
|-- dynamic-apps/
`-- run-app-elfloader/
```

```console

> $ cd run-app-elfloader/

Reviewer: First say what you want to do (explanations), then run the actual commands. Not the other way around, i.e. run commands, list explanations / rationale. Remove `$` sign, separate command from output.

Booting from ROM..Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
                  Atlas 0.13.1~21ce1acf
Hello World!

$ ./run.sh -k ../app-elfloader/build/elfloader_qemu-x86_64 -r ../dynamic-apps/sqlite3/ "/usr/bin/sqlite3"

SeaBIOS (version rel-1.16.2-0-gea1b7a073390-prebuilt.qemu.org)
Booting from ROM..Powered by
o.   .o       _ _               __ _
Oo   Oo  ___ (_) | __ __  __ _ ' _) :_
oO   oO ' _ `| | |/ /  _)' _` | |_|  _)
oOo oOO| | | | |   (| | | (_) |  _) :_
 OoOoO ._, ._:_:_,\_._,  .__,_:_, \___)
                  Atlas 0.13.1~21ce1acf
-- warning: cannot find home directory; cannot read ~/.sqliterc
SQLite version 3.22.0 2018-01-22 18:45:57
Enter ".help" for usage hints.
Connected to a transient in-memory database.
Use ".open FILENAME" to reopen on a persistent database.
sqlite> .exit
```

The last commands run the dynamic versions of `c++/helloworld` and `sqlite3` applications, the ones in [the `dynamic-apps` repository](https://github.com/unikraft/dynamic-apps).

> Similar to what you did before, run as many applications from the `dynamic-apps` repository using the `elfloader` image you have built and the `run.sh` script.

Reviewer: Replace this with the use of `run_app.sh` and ask to update `defaults`.
